### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by martini-source**
First draft of running cli.

### CODE REVIEW

File: main.go
Comment (mhutter) [inline main.go:160]: Maybe one big `switch` statement is enough, something like

```go
switch parsedType {
case "vshnpostgresql":
	serviceType = defaults.GetVSHNPostgreSQLDefault()
case "vshnredis":
	serviceType = defaults.GetVSHNRedisDefault()
// ...
}
```

Or bonus for storing all possible types in a `map[string]func() runtime.Object`, that would also allow to print all known types

```go
const TYPES := map[string]func() runtime.Object{
	"vshnpostgresql": defaults.GetVSHNPostgreSQLDefault,
	"vshnredis": defaults.GetVSHNRedisDefault,
}

// ...

fn, ok := TYPES[parsedType]
if !ok {
	return nil, fmt.Errorf(/* ... */)
}
// ...
```

**Review by mhutter — CHANGES_REQUESTED** [review]

File: go.mod
Comment (martini-source) [reply go.mod]: Done.

File: main.go
Comment (martini-source) [reply main.go:101]: Done

File: main.go
Comment (martini-source) [reply main.go:125]: Done

File: main.go
Comment (martini-source) [reply main.go:200]: I refactored the defaults to offer the string map of functions.
I think the separation of different providers and with that the switch(although a more lean one) in findServiceType() are easier to maintain and keep track off instead of one big map or switch.

File: main.go
Comment (martini-source) [reply main.go]: I haven't yet. @davidgubler mentioned he wanted to check if DeepCopy is needed at all.

File: main.go
Comment (davidgubler) [inline main.go:101]: I feel that this implements stuff that the "github.com/spf13/pflag" library would do for us. Is there any specific reason why we're not using that library?

File: main.go
Comment (davidgubler) [inline main.go:75]: My IDE doesn't like the cast to type "[2]string" even after I've switched to go 1.20, but it compiles on the command line. Do you know what's going on here?

File: main.go
Comment (davidgubler) [inline main.go:70]: I don't like this because this makes it impossible to use "=" in values, and I see no reason to disallow that.

I suggest just splitting the string at the first "=" and leaving everything after that intact. The strings.Split function is probably not the right one to use for that, according to the strings docs "strings.Cut" might be the way to go.

File: main.go
Comment (davidgubler) [inline main.go:167]: We should probably also support float and double.

File: main.go
Comment (davidgubler) [inline main.go:162]: This check should be changed to check of Int, Int8, Int16, Int32, Int64 ("Int" is equal to 2 and "Int64" is equal to 6, thus you can also do a range check 2-6).

Additionally we should probably support unsigned integers Uint, Unit8, Unit16, Uint32 and Uint64 (range 7-11). This needs an additional "if" statement because the setter function is called "SetUint".

File: main.go
Comment (davidgubler) [inline main.go:135]: I would not do this kind of differentiation between "case is correct" and "case differs, thus print an info message", instead I would always silently do case-insensitive matching that you do below (using fixStringCase). That would also allow us to get rid of the "strcase" library, too.

File: main.go
Comment (davidgubler) [inline main.go:187]: typo in "unsucsessful"

File: main.go
Comment (davidgubler) [inline main.go:130]: Why does this method operate on a copy of the object (I think) and not on a pointer? I think the "serviceType" argument should be a pointer and all the reflection magic should operate on that.

File: main.go
Comment (martini-source) [reply main.go:70]: Done

File: main.go
Comment (martini-source) [reply main.go]: Need for DeepCopyObject() resolved by commit 4a3c104ae3c7af510b9506e2c0b2f5f7830c13da

File: defaults/exoscalev1.go
Comment (martini-source) [reply defaults/exoscalev1.go]: Resolved by commit 2ae8d87084abb5c0d8933a806a4337a23f99b57d)

File: main.go
Comment (martini-source) [reply main.go:88]: Yes, since the flag itself represents the hierarchical "path" of the value to be set(I couldn't find how to make use of that with pflags), we don't have an fixed amount of valid input flags and we don't actually have to control behavior of appcat-cli through flags.

File: main.go
Comment (martini-source) [reply main.go:118]: Resolved see commit [958cd9e](https://github.com/vshn/appcat-cli/pull/1/commits/958cd9e3843de77a73d56d5e66c849b6ca6d4c00)

File: main.go
Comment (martini-source) [reply main.go:154]: See review comment below.

File: main.go
Comment (martini-source) [reply main.go]: Added proper checking and support for lnts Uints and Floats.
Resolved by commit [f91af1f](https://github.com/vshn/appcat-cli/pull/1/commits/f91af1fa086ab4d0727af11c659096d1946339fd)

File: main.go
Comment (davidgubler) [inline main.go:70]: This method of argument parsing is a bit wonky. I think it will fail for "--foo" "x=y" (key 'foo' with value 'x=y'), and probably also for "--foo" "--bar" (key 'foo' with value '--bar'). These are of course border cases and I don't mind leaving the code as it is for now (we can always improve things later), but the argument parsing should probably work more like this:
* Single pass over all arguments, have a map of all arguments
* Upon encountering "--(.*)=(.*)" put into map $1 as key and $2 as value
* Upon encountering "--(.*)" put into map $1 as key and next argument (unconditionally) as value (consume two arguments)
* Upon encountering anything else print warning

(the regexes are just for illustration, don't implement something like this with regexes)

Again, I'm OK with as it is for now, but I have to mention this for completeness.

File: main.go
Comment (davidgubler) [reply main.go:87]: Makes a lot of sense, thanks!

File: main.go
Comment (davidgubler) [reply main.go:118]: Thanks!

**Comment by davidgubler** [comment]
Most of the commits should probably be combined before merging this to master, but apart from that LGTM, thank you!

File: main.go
Comment (martini-source) [reply main.go:70]: Are there use cases with --foo --bar as key/value ? In the event of a missing value we would loose the ability to tell which key is missing it's value, which for longer parameter lists could be a headache.

Comment (davidgubler) [reply main.go:70]: I don't think it is worth discussing this at this point, and arguments can be made in many directions, I just wanted to point out that the current implementation doesn't parse all cases correctly.

Comment (martini-source) [reply main.go:70]: Alright, I added a fix for the value contains = case and added error for missing values.
Will do a proper refactor if we run into issues with the other edge case in the future.

Comment (davidgubler) [reply main.go:70]: :+1: Thank you!

**Comment by martini-source** [comment]
Squashed all the commits into one, I'll wait for @mhutter to resolve reviews and merge to master after.

**Review by mhutter — APPROVED** [review]

**Pull request merged**
